pylint/checkers/refactoring/refactoring_checker.py

```diff
@@ -1601,3 +1614,167 @@ def _check_return_at_the_end(self, node):
+                        ):
+                            continue
+
+                        last_definition_lineno = value.lookup(value.name)[1][-1].lineno
+                        if last_definition_lineno > node.lineno:
+                            # Ignore this subscript if it has been redefined after
+                            # the for loop. This checks for the line number using .lookup()
+                            # to get the line number where the iterating object was last
+                            # defined and compare that to the for loop's line number
+                            continue
+
+                        self.add_message(
+                            "unnecessary-dict-indexing",
+                            node=subscript,
+                            args=(node.target.elts[1].as_string()),
+                        )
+
+                    # Case where .items is assigned to single var (i.e., for item in d.items())
+                    elif isinstance(value, astroid.Subscript):
+                        # Note: We don't have to check for reassignment for this case as
+                        # tuples are immutable
+                        if (
+                            not isinstance(node.target, astroid.AssignName)
+                            or node.target.name != value.value.name
+                            or iterating_object_name != subscript.value.as_string()
+                        ):
+                            continue
+                        # check if subscripted by 0 (key)
+                        subscript_val = value.slice
+                        if isinstance(subscript_val, astroid.Index):
+                            subscript_val = subscript_val.value
+                        inferred = utils.safe_infer(subscript_val)
+                        if not isinstance(inferred, astroid.Const):
+                            continue
+                        inferred = cast(astroid.Const, inferred)
+                        if inferred.value != 0:
+                            continue
+                        self.add_message(
+                            "unnecessary-dict-indexing",
+                            node=subscript,
+                            args=("1".join(value.as_string().rsplit("0", maxsplit=1)),),
+                        )
+
+                    else:
+                        continue
+
+    def _check_unnecessary_dict_indexing_comprehension(
+        self, node: astroid.Comprehension
+    ) -> None:
+        if (
+            isinstance(node.iter, astroid.Call)
+            and isinstance(node.iter.func, astroid.Attribute)
+            and node.iter.func.attrname == "items"
+        ):
+            inferred = utils.safe_infer(node.iter.func)
+            if not isinstance(inferred, astroid.BoundMethod):
+                return
+            iterating_object_name = node.iter.as_string().rpartition(".items")[0]
```

Comment:
I'm wondering if some of it can be merged with the previous method? If I'm seeing this correctly, much of it is similar after all. Maybe one or two helper methods?

Reply:
Managed to merged into one single method :)

Reply:
see: 14c1c2fd